### **Telco Customer Churn Prediction - Model Comparison**

**Objective:** To develop a machine learning classification model that predicts customers who are likely to discontinue their telecom subscription services. The project aims to identify the key factors influencing customer churn through data cleaning, feature engineering, model training, and evaluation, enabling telecom companies to implement targeted customer retention strategies and reduce revenue loss.

**Guiding Principles for this Notebook:** 

+ Import Libraries
+ Load Datas
+ Compare

**Import Libraries**

In [1]:
import pandas as pd
import numpy as np

**Load datasets**

In [2]:
original_results = pd.read_csv("C:\\Users\\KOLADE\\OneDrive\\Documents\\AkoladeDSJourney\\Telco-Customer-Churn-Prediction\\reports\\results\\original_results.csv")
original = original_results.copy()

cw_results = pd.read_csv("C:\\Users\\KOLADE\\OneDrive\\Documents\\AkoladeDSJourney\\Telco-Customer-Churn-Prediction\\reports\\results\\cw_results.csv")
cw = cw_results.copy()

smote_results = pd.read_csv("C:\\Users\\KOLADE\\OneDrive\\Documents\\AkoladeDSJourney\\Telco-Customer-Churn-Prediction\\reports\\results\\smote_results.csv")
smote = smote_results.copy()

In [6]:
results = pd.concat([original, cw, smote], axis=0, ignore_index=True)

results.sort_values(by=['Recall', 'F1', 'ROC_AUC'], ascending=False, inplace=True)

display(results)

,Experiment,Model,Accuracy,Precision,Recall,F1,ROC_AUC
4,original + class weights,Logistic Regression,0.7539,0.5301,0.8293,0.6467,0.8550
8,SMOTE,Logistic Regression,0.7525,0.5288,0.8153,0.6415,0.8541
7,original + class weights,XGBoost,0.7728,0.5666,0.6969,0.6250,0.8330
11,SMOTE,XGBoost,0.7866,0.6059,0.6132,0.6095,0.8303
10,SMOTE,Random Forest,0.7761,0.5933,0.5592,0.5758,0.8225
0,original_train,Logistic Regression,0.8074,0.6765,0.5575,0.6113,0.8557
9,SMOTE,Decision Tree,0.7227,0.4905,0.5418,0.5149,0.6647
3,original_train,XGBoost,0.7937,0.6500,0.5209,0.5783,0.8321
2,original_train,Random Forest,0.7974,0.6698,0.5017,0.5737,0.8278
1,original_train,Decision Tree,0.7298,0.5027,0.4808,0.4915,0.6520


#### **Model Performance Summary**

Three imbalance-handling strategies were evaluated:

1. Original training data without intervention.
2. Original data with algorithm-level balancing techniques (`class_weight='balanced'` and `scale_pos_weight`).
3. SMOTE-based oversampling on the training data.

The results showed that imbalance-handling effectiveness varied across algorithms.

Logistic Regression achieved the strongest overall performance when combined with class weighting, producing the highest Recall (82.9%) and F1-score (64.7%) while maintaining a high ROC-AUC (85.5%). This makes it particularly suitable for churn prediction, where identifying customers likely to leave is more important than minimizing false positives (*Recall > Precision* - the usual trade-off).

XGBoost benefited substantially from `scale_pos_weight`, outperforming its SMOTE counterpart and demonstrating that algorithm-level methods can be more effective than synthetic oversampling for boosting algorithms.

Random Forest and Decision Tree models showed modest improvements under SMOTE, though their overall performance remained below that of Logistic Regression and XGBoost.

Overall, the findings suggest that algorithm-level imbalance handling is sufficient for moderately imbalanced datasets such as Telco Churn, and that synthetic oversampling should be validated empirically rather than assumed to improve performance.

>The next logical step is hyperparameter tuning, especially for:
>
> + Logistic Regression (*C, penalty*)
> + Random Forest (*max_depth, min_samples_leaf, n_estimators*)
> + XGBoost (*learning_rate, max_depth, subsample, n_estimators, gamma*)

**Next:** Telco Customer Churn Prediction - Hyperparameter tuning
> **notebook:** *`08_hyperparameter_tuning.ipynb`*